packages/components/src/modal/index.tsx

```diff
@@ -75,7 +103,13 @@ function UnforwardedModal(
 	const headingId = title
 		? `components-modal-header-${ instanceId }`
 		: aria.labelledby;
-	const focusOnMountRef = useFocusOnMount( focusOnMount );
+
+	// If focusOnMount is `firstElement`, Modals should ignore the `Close` button which is the first focusable element.
+	// Remap `true` to select the next focusable element instead.
+	const focusOnMountRef = useFocusOnMount(
```

Comment:
Forgive me if it's already mentioned above but it's a long PR and I'm not on my laptop 😅 .

Could we instead just assign the `focusOnMountRef` to the `<div>` with `childrenContainerRef` on L320 when `focusOnMount === 'firstElement'`? This way we avoid the `use-focusOnMount` API change and the awkward CSS class name selector etc.

Reply:
That would also change the focus behaviour when the `focusOnMount` prop is set to `true`, because the children container would gain focus instead of the dialog container — in my opinion, that behaviour is not correct as we'd like to focus the dialog component.

Reply:
> when `focusOnMount === 'firstElement'`

I mean only change the ref callback to be assigned to the children container div **if and only if** `foucsOnMount === 'firstElement'`. Is it still going to be incorrect?

Reply:
> > when `focusOnMount === 'firstElement'`
> 
> I mean only change the ref callback to be assigned to the children container div **if and only if** `foucsOnMount === 'firstElement'`. Is it still going to be incorrect?

I wouldn't know off the top of my head, but I guess we could try the approach keeping the current set of unit tests and see how it goes.

Reply:
Would this give us the flexibility to find the first focusable node _outside_ of the content wrapper if there's nothing appropriate within it? Feels like we still need the "global" context, and moving the `ref` would take that away.

Reply:
Unless I"m mistaken @andrewhayward is right. The tests assert that if no focusable node is found within the Modal contents then it should seek a node anyway within the Modal (including outside of the contents div - i.e. the one with the `childrenContainerRef` on it).

Reply:
IMHO, that's the responsibility of the consumer to render focusable elements in `children` if they explicitly pass `'firstElement'` for the `focusOnMount` prop. If there are no focusable elements in the content wrapper, then they should just pass `true` or `false` and roll their own focus logic if needed. It depends on how we want the consumers to use this component and how much we want to do for them. This behavior also matches what the doc says more closely:

> If the value `firstElement` is used then the component will attempt to place focus within the Modal's **contents**

It doesn't mention anything about a fallback outside the Modal's contents, and that might seem unexpected to some.

Reply:
Just noting that I believe not including finding elements outside the contents when passing `firstElement` might represent a minor regression. Though of course, changing where focus is placed is already a potentially breaking change...

Reply:
> > If the value `firstElement` is used then the component will attempt to place focus within the Modal's **contents**
> 
> It doesn't mention anything about a fallback outside the Modal's contents, and that might seem unexpected to some.

That's a good point, and I would be totally up for testing it out if were building a new component from scratch —  although I see how someone could argue that a close button would still be part of the modal's contents.

But as @getdave mentions, this would definitely introduce a breaking change (accessibility regression) for those modals where the only tabbable element is the close button, and where the consumer of the modal specified `focusOnMount="firstElement"`.

> Though of course, changing where focus is placed is already a potentially breaking change...

This is also a fair point, although IMO the regression listed above (ending up with no element being focused) is much worse than having a different element being focused.

---

In short, there's no perfect solution. But I feel like the current solution may represent a more sensible compromise for existing `Modal` consumers.

Having said that, if folks believe that the approach proposed by @kevin940726 is better / worth a shot, please feel free to do so.

Reply:
True! FWIW, I would prefer to add a new type like `firstContentElement` instead of patching `firstElement`. It's also clearer for folks that don't read the doc/changelog or have false (but fair) assumptions.

That said, I don't really have a strong preference here. I'm just sharing a suggestion, and I'm okay with whatever solution we end up with! ❤️

Reply:
> I would prefer to add a new type like `firstContentElement` instead of patching `firstElement`

That could also be an option, I wouldn't be opposed to that.

Reply:
Using `firstContentElement` would avoid regressions. I'm tempted to try that approach next week alongside the suggestion to conditionally move the `ref` when `firstContentElement` is set.

I knew this wouldn't be easy 😅

Reply:
You all are doing a great job! Happy to help in any way when I have the capacity! 💪

Reply:
@kevin940726 Alternative now available in https://github.com/WordPress/gutenberg/pull/54590